resourcemonitor: switch to ghw

ffromani opened the pull request:
We can use GHW facilities for everything bar the NUMA distances - which we use to estimate the costs.
So, we fork ghw and augment it to report NUMA distances; we are going to submit the patch to u/s asap.
Meantime, we consume the fork to further validate it.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
resourcemonitor: switch to ghw

We can use GHW facilities for everything bar the NUMA distances.
Thus, we switch to a fork which adds support for the distances
reporting, until upstream catches up.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
conflicts with https://github.com/swatisehgal/node-feature-discovery/pull/10

swatisehgal commented:
Can you rebase this please.

ffromani commented:
> Can you rebase this please.

needs rebase and (quite some) fixes to the tests. Will update ASAP.

ffromani commented:
/hold

ffromani commented:
/hold cancel

swatisehgal commented on pkg/resourcemonitor/noderesourcesaggregator.go:
```diff
@@ -41,30 +40,18 @@ type nodeResources struct {
-	}
-
-	dists, err := distances.NewDistancesFromSysfs(sysfsPath)
+	topo, err := ghw.Topology(ghw.WithChroot(sysfsRoot))
```
Reading through [this](https://github.com/jaypipes/ghw#overriding-the-root-mountpoint-ghw-uses), I've noticed that WithChroot is used to override root mountpoint and not sysfs like we are doing in this case. Because of this we will end up using` /host-sys/sys` as our `sysfs` and cpu resources will not computed/accounted correctly.

ffromani replied:
Right! Thanks for spotting this. This is a limitation (pretty fair however) of how ghw works. I think we can sidestep this problem changing the way we mount the host sysfs and procfs inside the NFD container. Would that work for you?
If so I'll add the patch to change the mountpoints in this PR.

swatisehgal replied:
The changes look good. But, I think we are missing one thing as we need to make sure that in call to WithChroot we override root mountpoint correctly, we need to update:

-  either the argument in the call to resourcemonitor.NewResourcesAggregator() in main so it receives `/host` as its root and not `/host/sys`  as in the current implementation we end up querying `/host/sys/sys` as our `sysfs` 

- or update when we parse the sysfs arg [here](https://github.com/swatisehgal/node-feature-discovery/blob/rte/cmd/nfd-topology-updater/main.go#L188)

ffromani committed:
nfd-topology-updater: manifests: fix /sys mount

Due to how ghw works, we need to mount the host sysfs
in a chroot friendly way. Nothing really changes besides
the mount paths inside the containers.

We need to repurpose the `--sysfs` option to be the chroot
 -not the full path on which sysfs is mounted.

Before:
--sysfs=/host-sysfs  -> sysfs root = /host-sysfs/

After:
--sysfs=/host        -> sysfs root = /host/sys/

(e.g. the "sys" part is implicitely and automatically added now - by GHW).

Unfortunately this is an API breakage, but since we didn't make any
backward compatibility promise yet, this is a breakage we can still
afford.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
Very good point. Updating as suggested.
